docker/algod/source/Dockerfile

```diff
@@ -0,0 +1,34 @@
+FROM golang:1.13.12
+
+ARG URL=https://github.com/algorand/go-algorand
+ARG BRANCH=master
+
+RUN echo "Installing from source. ${URL} -- ${BRANCH}"
+ENV BIN_DIR="$GOPATH/bin"
+
+ENV DEBIAN_FRONTEND noninteractive
+
+# Basic dependencies.
+ENV HOME /opt
+RUN apt-get update && apt-get install -y apt-utils curl git git-core bsdmainutils python3
+
+# Copy everything into the container.
+COPY . /tmp
+
+# Install algod binaries.
+RUN git clone --single-branch --branch "${BRANCH}" "${URL}" && \
```

Comment:
This isn't a functional issue, but it may be better to have this git clone be done by the up script, then use these docker steps to take advantage of docker caching better

```
COPY go-algorand/scripts/configure_dev.sh /tmp/go-algorand/scripts/configure_dev.sh
COPY go-algorand/scripts/configure_dev-deps.sh /tmp/go-algorand/scripts/configure_dev-deps.sh
RUN cd /tmp/go-algorand && \
         ./scripts/configure_dev.sh
COPY go-algorand /tmp/go-algorand
RUN make install
```

The same goes for the COPY . /tmp part too actually. When you add every file like that in the beginning, if any single file changes the entire docker build cache is reset. To take better advantage of it you can add individual files for processes that take a long time and then throw everything you need in at the end.

Reply:
I don't think caching helps much for this environment because we're building from source, and the source is constantly changing.

The only really cachable bit is libsodium, and in theory it could be cached by building master, then adding a different remote / checking out the target branch. I think that would add too much complexity for the speedup we'd see.

Reply:
The docker build process actually accounts for that though. The hash computed for these copy steps is computed based on the contents that are copied over. In this case if `configure_dev.sh` or `configure_dev-deps.sh` change, the cache will be invalidated and and subsequent steps will be run, but if they don't the cached execution of those scripts can be used. This way you can save time by only running those scripts on changes. I'm with you though that taking advantage of this may be more work than it's worth for now, since our folder structure isn't really set up well for it.

Reply:
The source only exists in the container, so `git clone` needs to run in order to update these files